File: 2023/day07_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import itertools
from typing import Any

type Hand = tuple[str, str, int]


def solution(content: list[str]) -> Any:
    return sum(
        index * bid
        for index, (_, _, bid) in enumerate(
            sorted(
                sorted((parse(line) for line in content), key=sort_by_card),
                key=sort_by_type,
            ),
            start=1,
        )
    )


def sort_by_card(hand: Hand):
    card, _, _ = hand
    return ["23456789TJQKA".index(c) for c in card]


def sort_by_type(hand: Hand):
    _, tp, _ = hand
    if tp == "FiveOfKind":
        return 6
    if tp == "FourOfKind":
        return 5
    if tp == "FullHouse":
        return 4
    if tp == "ThreeOfKind":
        return 3
    if tp == "TwoPair":
        return 2
    if tp == "OnePair":
        return 1
    return 0


def parse_type(cards: str):
    signature = tuple(sorted(len(list(g)) for _, g in itertools.groupby(sorted(cards))))
    if (5,) == signature:
        return "FiveOfKind"
    if (1, 4) == signature:
        return "FourOfKind"
    if (2, 3) == signature:
        return "FullHouse"
    if (1, 1, 3) == signature:
        return "ThreeOfKind"
    if (1, 2, 2) == signature:
        return "TwoPair"
    if (1, 1, 1, 2) == signature:
        return "OnePair"
    return "HighCard"


def parse(line: str) -> Hand:
    cards, bid = line.split(" ")
    return (cards, parse_type(cards), int(bid))


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2023/day07-1.ts
type Poker =
  | "HighCard"
  | "OnePair"
  | "TwoPair"
  | "ThreeOfKind"
  | "FullHouse"
  | "FourOfKind"
  | "FiveOfKind";

type Hand = { cards: string; type: Poker; bid: number };

function solution(content: string) {
  return content
    .split("\n")
    .map(parse)
    .toSorted(compareCards)
    .toSorted(compareType)
    .map((hand, index) => hand.bid * (index + 1))
    .reduce((acc, cur) => acc + cur, 0);
}

function compareType(a: Hand, b: Hand) {
  return (
    [
      "HighCard",
      "OnePair",
      "TwoPair",
      "ThreeOfKind",
      "FullHouse",
      "FourOfKind",
      "FiveOfKind",
    ].indexOf(a.type) -
    [
      "HighCard",
      "OnePair",
      "TwoPair",
      "ThreeOfKind",
      "FullHouse",
      "FourOfKind",
      "FiveOfKind",
    ].indexOf(b.type)
  );
}

function compareCards(a: Hand, b: Hand): number {
  return (
    a.cards
      .split("")
      .map(
        (_, index) =>
          "23456789TJQKA".indexOf(a.cards.at(index)!) -
          "23456789TJQKA".indexOf(b.cards.at(index)!)
      )
      .find((diff) => diff !== 0) ?? 0
  );
}

function parseType(cards: string): Poker {
  const signature = Object.values(
    cards.split("").reduce((acc, cur) => {
      if (acc[cur]) return { ...acc, [cur]: acc[cur] + 1 };
      return { ...acc, [cur]: 1 };
    }, {} as { [k: string]: number })
  )
    .toSorted()
    .join(",");

  if (signature === "5") return "FiveOfKind";
  if (signature === "1,4") return "FourOfKind";
  if (signature === "2,3") return "FullHouse";
  if (signature === "1,1,3") return "ThreeOfKind";
  if (signature === "1,2,2") return "TwoPair";
  if (signature === "1,1,1,2") return "OnePair";
  return "HighCard";
}

function parse(line: string): Hand {
  const [cards, bid] = line.split(" ");
  return {
    cards,
    type: parseType(cards),
    bid: Number(bid),
  };
}

Deno.readTextFile("day07.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
